### ISSUE TITLE
Failover for celerybeat

### ISSUE THREAD

**Issue by kerwin** [celery/celery#1495]
I want to make schedule process run in multiple nodes to improve availability, for reducing single point issue.

**Comment by thedrow** [comment]
Related to #251.

**Comment by ask** [comment]
Closing this, as we don't have the resources to complete this task.

**Comment by magmax** [comment]
Sorry, but I do not understand this.
The proposed solution relies on RabbitMQ exclusive queues, what works as a distributed Mutex used to have a semaphore during the bootstrap. So, testing this is just like testing RabbitMQ.
In addition, this request would allow HA for Celery beat.

This has just a drawback: if the current beat server sends a beat and immediately dies, the semaphore-awaiting new beat server will send the beat again, no worrying about the beat time period. This should be another problem, that can be solved by avoiding to store anything in localhost, but this should be another issue, in my opinion.

We have had this patch working in production for a couple of years without any problem.

Anyways, I think closing an issue about HA just because there are no resources is not a good deal.

**Comment by ask** [comment]
If there's a pull request for the issue then that will stay open.

I think I have looked into the pull request you refer to, using exclusive queues, but I have not
yet had the time to verify that it works.   I was not aware that they are considered a sound implementation of a distributed mutex?  We would need to run a stress test on this implementation, to make sure (at the very least serious) edge cases are covered.

That you have been using this in production for years is certainly a very sound data point.  Have you tested it under stress? Killing nodes in different scenarios seeing how it recovers?
E.g. one thing I'm concerned about is that beat, by nature of being a scheduler, is dependent on a
single clock source to be consistent.  If there's failover it means there are multiple clocks, which means that periodic tasks can be sent in duplicate.  We need to figure out, what in a best case scenario with NTP synced clocks, and a maximum clock skew the maximum precision a periodic task can have before duplicates are sent.

The problem is that we have so many issues, but nobody to work on them.  Many of them we don't even know if is still an issue, or maybe if it's already fixed in master for the 4.0 release.  If the only progress is dead silence, then we can just as well close it.  In this case I knew there's at least one open pull request attempting to solve the issue.

I really would like to have failover for Beat, and it's definitely possible,  but we don't have any resources to put into it.  The only time I have is the weekend, but I also have to pay the rent and have a son to raise, so I have closed most of the feature request threads. 

I'm also open to reopening any of the issues that I have closed, you all have the ability to influence this!

**Comment by magmax** [comment]
Thank you for the explanation. It sounds really coherent.

**Comment by ask** [comment]
Had I time for a coherent responses, I would have spent that time more wisely.  In hindsight I should have taken the inevitable snarky response into account.

There are actionable steps here:

1) Describe to me how you tested the failover in different scenarios (over real network, unreachable, unresponsive) 
2) Describe to me when the implementation can result in duplicated messages.

From then we can proceed, as I simply don't have the incentive to investigate this myself at this moment.

**Comment by ask** [comment]
Actually, I considered it and the pull request is now closed.  We don't have time to bother with peoples snide remarks in open source projects.  I'm sorry I didn't review it earlier, but we are volunteers working for free, and this is the very reason why I've spent less time on the project and why we are at this point where I have to close 200 issues.  Maybe I can expect backlash for this decision, but I have literally nothing to lose, so go ahead.

**Comment by ant31** [comment]
The problem of distributed cron is complex and can be achieved via an external project.
If anyone wants to open a new repo to work on it, please share the link.

**Comment by magmax** [comment]
@ask, ok to all, but I want to answer your questions.

The implementation is just a semaphore before starting. I do not remember how I tested it, but maybe I launched a couple of beats and killed their connections from the RabbitMQ admin, trying different scenarios.

This may be a way to test it: creating two VM with Vagrant, one of them skewed some ms and with a random action just before sending the beat and just after: to tell Rabbit to disconnect a random connection. Running this for a long time you should be sure enough to trust it. Docker is not enough since it can not manage different clocks.

Duplicated messages: as long as the beat stores data in localhost, there is no way to know when the last beat happened, so the starting beat server cannot calculate the next execution. So, whenever the active beat server disconnect, the passive beat server can duplicate the beat.

I really respect you, @ask, because both Kombu and Celery are great tools, and I understand your decission. But current behaviour with two beat servers is to have duplicated beats and with just one the system has a Single Point of Failure (SPoF). With this semaphore, you would have a duplicated message if the active beat server dies, but avoids the SPoF. I think it is a big advantage, and whoever uses it must decide if can deal with duplicated messages from time to time.

Anyway, this is a tool to manage distributed systems, and distributed systems should deal with duplicated messages as a rule. But this should be the developer responsability.

Thank you, again, for such a tool and your time with this issue.

**Comment by ze-phyr-us** [comment]
https://github.com/aphyr/jepsen should be just the tool for stress-testing distributed systems. It can simulate clock skew among other things. The author tested a lot of popular systems with it: https://aphyr.com/tags/jepsen

**Comment by ddevlin** [comment]
For anyone else who ends up here while searching for a distributed / auto-scaling friendly celery beat, and is happy to use Redis as a backend; I tried both BeatCop and single-beat mentioned above, but ultimately picked [RedBeat](https://github.com/sibson/redbeat).

**Comment by matusvalo** [comment]
Hi all,

failover of celery beat can be implemented without any code change. It can be implemented using consul cluster[1], more precisely consul lock feature [2]. Basically, for having failover of celery beat you need to:
1. deploy consul cluster. At least 3 nodes are required. Deployment is pretty easy [3]
2. celery beat scheduler with central storage - e.g. django-celery-beat or other custom solution (Redis etc.)
3. celery beat service needs to run on at least two nodes under consul lock:
```
consul lock KEY celery -A proj beat -l info --scheduler django_celery_beat.schedulers:DatabaseScheduler
```

After implementing this three steps consul is guarantying that only one instance will be running at once. Consul is also responsible for starting "shadowing" instance in case when running one is failed/crashed. Regarding consul itself, it is highly available solution - see [4]

I hope that this helps. I am personally using consul for multiple scenarios but also I am using this failover feature.

[1] https://www.consul.io/
[2] https://www.consul.io/docs/commands/lock.html
[3] https://www.consul.io/docs/guides/bootstrapping.html
[4] https://www.consul.io/docs/internals/consensus.html#deployment-table

**Comment by auvipy** [comment]
thanks for sharing your views

**Comment by jonashaag** [comment]
FWIW, we've solved this by starting multiple https://github.com/ybrs/single-beat wrapped beats. single-beat makes sure only one beat process is running at the same time.

You could also use uWSGI's [legion mode](http://uwsgi-docs.readthedocs.io/en/latest/AttachingDaemons.html#legion-support).
